core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -242,64 +226,114 @@ void forceRefresh() {
    * value of currentInstanceData and schedules the next refresh shortly before the information
    * would expire.
    */
-  private InstanceData performRefresh() throws InterruptedException, ExecutionException {
+  @VisibleForTesting
+  InstanceData performRefresh() throws InterruptedException, ExecutionException {
     // To avoid unreasonable SQL Admin API usage, use a rate limit to throttle our usage.
     forcedRenewRateLimiter.acquirePermit();
 
-    ListenableFuture<InstanceData> refreshFuture;
+    ListenableFuture<InstanceData> refreshFuture = fetchInstanceData();
+
+    try {
+      InstanceData data = refreshFuture.get();
+
+      synchronized (instanceDataGuard) {
 
+        // update currentInstanceData with the most recent results
+        currentInstanceData = Futures.immediateFuture(data);
+
+        // schedule a replacement before the SSLContext expires;
+        nextInstanceData =
+            executor.schedule(
+                () -> performRefresh(),
+                new RefreshCalculator()
+                    .calculateSecondsUntilNextRefresh(
+                        Instant.now(), data.getExpiration().toInstant()),
+                TimeUnit.SECONDS);
+      }
+
+      return data;
+    } catch (Exception e) {
+      synchronized (instanceDataGuard) {
+        nextInstanceData = executor.submit(() -> performRefresh());
+      }
+      throw e;
+    }
+  }
+
+  @VisibleForTesting
+  public ListenableFuture<InstanceData> getCurrentInstanceData() {
+    synchronized (instanceDataGuard) {
+      return currentInstanceData;
+    }
+  }
+
+  @VisibleForTesting
+  ListenableFuture<InstanceData> getNextInstanceData() {
+    synchronized (instanceDataGuard) {
+      return nextInstanceData;
+    }
+  }
+
+  private ListenableFuture<InstanceData> fetchInstanceData() {
```

Comment:
I moved this function from `SqlAdminApiFetcher` so that all of the Future and Retry logic are contained in `CloudSqlInstance`. This leaves `SqlAdminApiFetcher` to strictly do API access and transformations.

Reply:
I think this is a step away from where we want to go. In my mind, we want the thing that gets the data from the API to encapsulate the how (i.e., two API calls instead of one) and instead be a one-method interface. This makes testing easier and using the interface easier. By moving this out we now have to coordinate the two API calls in a class that just wants the result of those calls: a certificate and an IP address.

Reply:
We should be able to just return an `InstanceData` without a future wrapping it. That seems like a good PR on its own honestly.